homework/index.js

```diff
@@ -1,54 +1,81 @@
 'use strict';
 
-{
-  function fetchJSON(url, cb) {
-    const xhr = new XMLHttpRequest();
-    xhr.open('GET', url);
-    xhr.responseType = 'json';
-    xhr.onload = () => {
-      if (xhr.status >= 200 && xhr.status <= 299) {
-        cb(null, xhr.response);
-      } else {
-        cb(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
-      }
-    };
-    xhr.onerror = () => cb(new Error('Network request failed'));
-    xhr.send();
-  }
+function fetchJSON(url, cb) {
+  const xhr = new XMLHttpRequest();
+  xhr.open('GET', url);
+  xhr.responseType = 'json';
+  xhr.onload = () => {
+    if (xhr.status >= 200 && xhr.status <= 299) {
+      cb(null, xhr.response);
+    } else {
+      cb(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
+    }
+  };
+  xhr.onerror = () => cb(new Error('Network request failed'));
+  xhr.send();
+}
 
-  function createAndAppend(name, parent, options = {}) {
-    const elem = document.createElement(name);
-    parent.appendChild(elem);
-    Object.entries(options).forEach(([key, value]) => {
-      if (key === 'text') {
-        elem.textContent = value;
-      } else {
-        elem.setAttribute(key, value);
-      }
-    });
-    return elem;
-  }
+function createAndAppend(name, parent, options = {}) {
+  const elem = document.createElement(name);
+  parent.appendChild(elem);
+  Object.entries(options).forEach(([key, value]) => {
+    if (key === 'text') {
+      elem.textContent = value;
+    } else {
+      elem.setAttribute(key, value);
+    }
+  });
+  return elem;
+}
 
-  function renderRepoDetails(repo, ul) {
-    createAndAppend('li', ul, { text: repo.name });
-  }
+function addRow(property, key, parent, link) {
+  const tr = createAndAppend('tr', parent);
+  createAndAppend('th', tr, {
+    text: `${property} :`,
+  });
 
-  function main(url) {
-    fetchJSON(url, (err, repos) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, {
-          text: err.message,
-          class: 'alert-error',
-        });
-        return;
-      }
-      const ul = createAndAppend('ul', root);
-      repos.forEach(repo => renderRepoDetails(repo, ul));
+  if (link) {
+    const td = createAndAppend('td', tr);
+    createAndAppend('a', td, {
+      href: link,
+      target: '_blank',
+      text: key,
     });
+  } else {
+    createAndAppend('td', tr, { text: key });
   }
+}
```

Comment:
You have understood well how this function can make your code DRY. Note however how this function includes code to deal with the exception for the first row, which should contain a link instead of simple text. A better strategy would be to let the caller of this function deal with the exception so that this function can be simplified to:.
```js
function addRow(property, key, parent) {
  const tr = createAndAppend('tr', parent);
  createAndAppend('th', tr, {
    text: `${property} :`,
  });
  createAndAppend('td', tr, { text: key });
  return tr;
}
```

Note that in the last line we are returning the created `tr` so that the caller of this function can pick it up.

Now, in your `renderRepoDetails()` function, you can deal with the first row as follows:

```js
const firstRow = addRow('Repository', '', table);
createAndAppend('a', firstRow.lastChild, {
  href: repo.html_url,
  target: '_blank',
  text: repo.name,
});
addRow('Description', ...);
```

Note that we pass an empty string for the `key` parameter and save the return value of the function into the `firstRow` variable. Then we add the link to the last child of the `tr`, which is, of course, the `td` where the link should go.